### PR TITLE
Tweak portfolio UI and add option to show hidden NFTs

### PR DESCRIPTION
**Pull request by simoarpe**
## Resolves https://github.com/brave/brave-browser/issues/32545

Screenshot_1693913055

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
- Open Brave Wallet and navigate to "NFTs" section by tapping on the tab at the top
- Observe NFTs and Hidden sections are shown
- Add a new NFT
- Observe the NFT is shown in the NFTs section
- Tap on "Edit Visible NFTs"
- Deselect the NFT and tap on "Save account watchlist"
- Observe the NFT is shown in the Hidden section

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/NftGridFragment.java [hunk 1]
```diff
@@ -384,7 +445,7 @@ private void recordP3AView() {
         if (activity instanceof BraveWalletActivity) {
             BraveWalletActivity walletActivity = (BraveWalletActivity) activity;
             BraveWalletP3a walletP3A = walletActivity.getBraveWalletP3A();
-            if (walletP3A != null) {
+            if (walletP3A != null && mNftDataModels != null) {
```
Comment (simoarpe) [inline android/java/org/chromium/chrome/browser/crypto_wallet/fragments/NftGridFragment.java:448]: Improved check as the wallet was crashing here.

**Review by mkarolin — APPROVED** [review]
`strings`++

**Review by Pavneet-Sing — APPROVED** [review]
LGTM, added some comments to improve performance. (can be addressed later if suitable).

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/NftGridFragment.java [hunk 2]
```diff
@@ -348,15 +408,16 @@ private void updateNftGrid() {
...
                     if (!AndroidUtils.canUpdateFragmentUi(this)) return;
                     mPortfolioHelper = portfolioHelper;
                     List<BlockchainToken> nfts = mPortfolioHelper.getUserAssets();
-                    LiveDataUtil.observeOnce(getNetworkModel().mCryptoNetworks, networkInfos -> {
-                        mPortfolioModel.prepareNftListMetaData(
-                                nfts, networkInfos, mPortfolioHelper);
-                    });
+                    List<BlockchainToken> hiddenNfts = mPortfolioHelper.getHiddenAssets();
+                    LiveDataUtil.observeOnce(getNetworkModel().mCryptoNetworks,
+                            networkList
```
[the remark this answers is not in the document]
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/NftGridFragment.java:418]: Oh, right! Nice catch, thank you!

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/NftGridFragment.java [hunk 3]
```diff
@@ -348,15 +408,16 @@ private void updateNftGrid() {
...
+                            -> mPortfolioModel.prepareNftListMetaData(
+                                    nfts, hiddenNfts, networkList, mPortfolioHelper));
```
[the remark this answers is not in the document]
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/NftGridFragment.java]: 👍

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/PortfolioHelper.java [hunk 4]
```diff
@@ -139,6 +144,61 @@ public void fetchAssetsAndDetails(
...
+
+                        if (userAssetsCount.incrementAndGet() == totalNetworks) {
+                            mUserAssets.sort(Comparator.comparing(
+                                    token -> mAssertSortPriorityPerCoinIndex.get(token.chainId)));
+                            AtomicInteger allAssets = new AtomicInteger();
+                            final List<BlockchainToken> assets = new ArrayList<>();
+                            for (NetworkInfo selectedNetwork : mSelectedNetworks) {
+                                TokenUtils.getUserOrAllTokensFiltered(braveWalletService,
+                                        blockchainRegistry, selectedNetwork, selectedNetwork.coin,
+                                        TokenUtils.TokenType.NFTS, false, tokens -> {
```
[the remark this answers is not in the document]
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/PortfolioHelper.java]: 👍

**Pull request merged**